homeassistant/components/device_tracker/icloud.py

```diff
@@ -1,100 +1,497 @@
+                           'reset_account_icloud', reset_account_icloud)
 
+    def setinterval(call):
+        """Call the update function of an icloud account."""
+        accountname = call.data.get(ATTR_ACCOUNTNAME)
+        interval = call.data.get(ATTR_INTERVAL)
+        devicename = call.data.get(ATTR_DEVICENAME)
+        if accountname is None:
+            for account in ICLOUDTRACKERS:
+                ICLOUDTRACKERS[account].setinterval(interval, devicename)
+        elif accountname in ICLOUDTRACKERS:
+            ICLOUDTRACKERS[accountname].setinterval(interval, devicename)
+
+    hass.services.register(DOMAIN,
+                           'setinterval', setinterval)
+
+    # Tells the bootstrapper that the component was successfully initialized
     return True
+
+
+class Icloud(object):  # pylint: disable=too-many-instance-attributes
+    """Represent an icloud account in Home Assistant."""
+
+    def __init__(self, hass, username, password, cookiedirectory, name,
+                 ignored_devices, googletraveltime, see):
+        """Initialize an iCloud account."""
+        # pylint: disable=too-many-arguments,too-many-branches
+        # pylint: disable=too-many-statements,too-many-locals
+        self.hass = hass
+        self.username = username
+        self.password = password
+        self.cookiedir = cookiedirectory
+        self.api = None
+        self.accountname = name
+        self.devices = {}
+        self._ignored_devices = ignored_devices
+        self._ignored_identifiers = {}
+        self.googletraveltime = googletraveltime
+        self._overridestates = {}
+        self._intervals = {}
+        self.see = see
+
+        self._trusted_device = None
+        self._verification_code = None
+
+        self._attrs = {}
+        self._attrs[ATTR_ACCOUNTNAME] = name
+
+        if self.username is None or self.password is None:
+            _LOGGER.error('Must specify a username and password')
+        else:
+            from pyicloud import PyiCloudService
+            from pyicloud.exceptions import PyiCloudFailedLoginException
+            try:
+                self.api = PyiCloudService(
+                    self.username, self.password,
+                    cookie_directory=self.cookiedir, verify=True)
+                for device in self.api.devices:
+                    status = device.status(DEVICESTATUSSET)
+                    devicename = slugify(status['name'].replace(' ', '', 99))
+                    if (devicename not in self.devices and
+                            devicename not in self._ignored_devices):
+                        self.devices[devicename] = device
+                        self._intervals[devicename] = 1
+                        self._overridestates[devicename] = None
+                    elif devicename in self._ignored_devices:
+                        self._ignored_identifiers[devicename] = device
+
+            except PyiCloudFailedLoginException as error:
```

Comment:
Extract everything in this `try:` block to reuse `reset_account_icloud` to remove duplication